### ISSUE TITLE
Placeholder / Unobtrusive Validation

### ISSUE THREAD

**Issue by jrichview** [RobinHerbots/Inputmask#2440]
I have a date field that is not required. When the user submits the form having NEVER visited the field, the placeholder text causes the date validation to fail. Obviously the string "MM/DD/YYYY" is not a valid date. 

I had expected the placeholder text to be just cosmetic, and the actual value of the input to be an empty string.

Here is how I am applying it:
` .inputmask({ "mask": "99/99/9999", placeholder: "MM/DD/YYYY", clearMaskOnLostFocus: false });`

We want the mask to be applied if the user inputs anything. We don't want just digits. But if they have entered nothing in the field, we should be getting an empty input value.

**Comment by jrichview** [comment]
Seems I have to caveat my issue.  The issue occurs after the form is submitted and fails server-side validation. The form is reloaded, the inputmask is applied during the document.ready() event and the validation complains that "MM/DD/YYYY" is not a valid date.

**Comment by jrichview** [comment]
In the MVC Controller's post method, the model's value for that field is null, which is fine (it's a nullable datetime). But the ModelState has an error for that field: 

![image](https://user-images.githubusercontent.com/2236908/102931620-b47a4200-446c-11eb-92d9-a069f212ed5d.png)

**Comment by jrichview** [comment]
UPDATE: As a workaround I've tried hooking the unobtrusive validation to add a normalizer for that field to allow me to strip out the placeholder text (code below).  What I found was that the value being passed to the validation code is an empty string. 

My conclusion is that this must be a problem with MVC submitting the form, not with the validation itself.  It is somehow grabbing the placeholder text and trying to convert that to nullable datetime.


        $("#myForm").validate({
            rules: {
                DateOfBirthSearch: {
                    normalizer: function (value) {
                        debugger;
                        return value.replace("MM/DD/YYYY", "");
                    }
                }
            }

**Comment by jrichview** [comment]
It seems that this plugin is the source of the issue. If I create a dummy input with a placeholder and don't enter anything into the field, it DOES NOT submit the placeholder text when the form submits.  So this seems to be a bug caused by inputmask. 

I seriously doubt most developers would want the input mask text submitted if nothing has been entered by the user, but at minimum this should be an option in the plugin which defaults to NOT submitting the inputmask text.

**Comment by RobinHerbots** [comment]
Hi @jrichview , 

Can you provide some more information.  
Have you tried the removemaskonsubmit option? https://github.com/RobinHerbots/Inputmask#removemaskonsubmit

On which platform or browser are you testing?

**Comment by jrichview** [comment]
Yes, but again we don't _want_ it unmasked. Bear in mind the mask and the placeholder are not the same thing. If the user enters a date in a date masked field we want them not to have to enter anything but digits, and that works perfectly. In that situation we don't want the mask removed. But if they don't input anything in the field at all, we don't want the placeholder value submitted as if the user had entered it. And that is what we're seeing.  

In a date field which is a required input, it's not a concern, but when the field is optional, it's a problem. We currently have some kludge code to look for the user having submitted the string "MM/DD/YYYY" and removing it.

This is on Windows 10, tested with IE, Edge, Chrome, and Firefox.

**Comment by RobinHerbots** [comment]
@jrichview , 

Which version do you use of the inputmask?

**Comment by jrichview** [comment]
5.05

**Comment by RobinHerbots** [comment]
Can you have a try with the latest beta version?  5.0.6-beta.25

**Comment by jrichview** [comment]
Ok will do as quickly as I can get to it and will let you know. Thanks!

**Comment by jrichview** [comment]
Didn't see a Beta package available in nuget so I applied the update manually and re-tested. Got the same issue. Here's a pic of the form to help explain. On this form user must enter SSN and either last name or DOB, then click search. If they have not yet entered the DOB field, then the value of the placeholder gets posted "MM/DD/YYYY" when it should be "".

![image](https://user-images.githubusercontent.com/2236908/103822235-b3850b00-503d-11eb-9557-bc583697536b.png)

**Comment by jrichview** [comment]
This has become a significant issue in one of our applications. We have search forms with quite a few optional date fields. When the user submits the form it is typical for the controller method to run code like the following:
`            
if (ModelState.IsValid)
{
   ...do something with the submitted model
}
`
Unfortunately, IsValid always returns false for date fields that had nothing entered for them because what was submitted is the placeholder string "MM/DD/YYYY" and we have to write kludge code for every date field to check for that problem and remove the error.

**Comment by RobinHerbots** [comment]
Why do you use clearMaskOnLostFocus: false?

**Comment by jrichview** [comment]
Setting clearMaskOnLostFocus to true avoids the problem because it removes the placeholder text. But we actually want the placeholder text to display for empty fields. 

I mean, that IS the design of placeholder text in the first place.  If you put the following on a form:

                              `                        <input placeholder="placeholder test">`

...you will see that placeholder text until the user enters a value regardless of where the keyboard focus is.

**Comment by RobinHerbots** [comment]
regardless of where the keyboard focus is.

Try setting the showMaskOnFocus: false & showMaskOnHover: false

**Comment by jrichview** [comment]
I tested with the following options:

`    $("input.date-picker").inputmask({
        "mask": "99/99/9999",
        placeholder: "MM/DD/YYYY",
        clearMaskOnLostFocus: false,
        autoUnmask: false,
        showMaskOnFocus: false,
        showMaskOnHover: false
    });
`

Same results.  The submitted data model's nullable DateTime fields were null (which is correct), but the ModelState had an error for each field. 

![image](https://user-images.githubusercontent.com/2236908/127164495-829dd8d1-1f48-456e-bf5b-13944c0ab1e3.png)

**Comment by jrichview** [comment]
It should be possible to reproduce this with a very minimal/default MVC project that has a nullable date time field on a submitted CREATE form.

**Comment by RobinHerbots** [comment]
Use the same options with the clearMaskOnLostFocus: true

**Comment by jrichview** [comment]
That works fine. Any time that value is set to true it sidesteps the problem because the placeholder text is removed.

**Comment by RobinHerbots** [comment]
@jrichview ,

```HTML
 <input id="test1" name="test1" autocomplete="name" maxlength="10" placeholder="Hallo Placeholder"/>
```
```
	Inputmask({
			"mask": "99/99/9999",
			placeholder: "MM/DD/YYYY",
			clearMaskOnLostFocus: true,
			autoUnmask: false,
			showMaskOnFocus: false,
			showMaskOnHover: false
		}).mask("test1");
```
Submitted formdata
 test1: 
test2: 
test3: 
test4: 
test5:

**Comment by jrichview** [comment]
Sorry, not sure what you're saying here?

**Comment by RobinHerbots** [comment]
@jrichview ,

That was jsut an example, I didn't yet see your asnwer.  Without the clearmaskonlostfocus, once the mask is set for some reason, focus, initial focus, entering something and deleting it again, it stays there, unless you set this option to its default, which will remove the mask (or placeholder in context of the inputmask) and show again the HTML placeholder.

**Comment by jrichview** [comment]
The following code always demonstrates the problem:

`    $("input.date-picker").inputmask({
        "mask": "99/99/9999",
        placeholder: "MM/DD/YYYY",
        clearMaskOnLostFocus: false
    });
`
And setting clearMaskOnLostFocus: true avoids the problem.  Unfortunately, our product management team were asking for the form/field behavior you get when you set it to true.

**This seems to be a specific conflict with MVC's model binding.** I am not sure whether it could be related to jquery unobtrusive validation (used by default in MVC), but I don't think so because I have tried hooking into the unobtrusive validation process and found that the value it was trying to validate was empty, not the placeholder text.

**Comment by RobinHerbots** [comment]
That is not really a problem.

The  form submits the current value of the input to the server.  When the mask is displayed (MM/DD/YYYY) that is text, so it gets submitted. The clearMaskOnLostFocus prevents submitting the mask value (emptymask) to the server.  That is what this option is for and its default behavior is to remove the mask/placeholder wxhen loosing focus.  Isn'"t that exactly what you want?

**Comment by jrichview** [comment]
An input's placeholder text should not be submitted or validated. It's only designed to guide the user what to input in the field.

I am confident this is a problem specific to this plugin.  I have tried the plugin both with a plain input field and with a Kendo datepicker and I get the same problem either way.  If I manually apply the HTML attribute placeholder="Placeholder Text" to an input or to a Kendo datepicker, I do not get a validation error.

If it would be helpful to you, I can create a very simple MVC project that demonstrates the issue.

I should clarify this, too: The model is not submitted to the controller method with the placeholder text.  The field's data type is nullable DateTime and it is posted as a null value.  However, that is because MVC's model binder tries to convert the placeholder text to a DateTime and that, of course, fails. When it fails, an error is added to the "ModelState" for that post, and that is the essence of the problem.

So what I am saying is that the conflict is with MVC's model binder.

**Comment by RobinHerbots** [comment]
@jrichview ,

I think you are confusing, the html placeholder with the maskplaceholder.
Anyway, what holds you from using these options
clearMaskOnLostFocus: true,
			autoUnmask: false,
			showMaskOnFocus: false,
			showMaskOnHover: false

It would be nice to recieve an example project, to clairify your exact needs.  Can you also decribe the exact expected behavior.  Initial visual on the screen.  Changes on focus, entering text etc, ....

**Comment by jrichview** [comment]
We're setting "clearMaskOnLostFocus" to false because we would like the same behavior as the plain old HTML input's placeholder attribute. That is, we want the date fields to display the text "MM/DD/YYYY" until the user actually starts entering something in the fields.  The text would be displayed when they first see the form, and tabbing through the form fields without entering anything would not change the placeholder text. Only typing a value in the input should remove the placeholder text. Here is an example of me entering a date (works GREAT!):  
![image](https://user-images.githubusercontent.com/2236908/127177957-c6d41718-b071-47d0-8389-815ee59c8aac.png)

This all works beautifully, and inputting a date works exactly how we want it to work using your input mask plugin.  EXCEPT:

The only problem we're having is the plugin's conflict with MVC binder which causes a false validation error on every date field which is NOT required and which has had no value entered.

I will put together a demo project and link it later today.

**Comment by eduardo-mior** [comment]
@jrichview 
Just a silly question.
Do users use this date filter? I'm working a lot with UI/UX lately, and as much as we provide dozens of filters, sometimes users end up not using any of them.
I'm working on applications with hundreds of users, our team uses Hotjar and Clarity to monitor what customers do. We remove a lot of useless stuff from our applications because users just don't use it. We removed a lot of filters that we found useful but none of the users have ever used them.
I just asked this question because maybe the simplest way to solve it is to remove the filter.
I'm not trying to be ignorant, I'm just trying to help.

**Comment by jrichview** [comment]
Eduardo: This is a problem with ANY date field which is not required on a form. And yes. They DO use these search filters.  We're making an updated version of an existing application, so we know this.

Robin: Please have a look at this:  [https://github.com/jrichview/InputMaskValidationErrorDemo](https://github.com/jrichview/InputMaskValidationErrorDemo)

Sorry there are so many files in a default MVC project.  The key files to look at are Views\Home\Index.cshtml and Controllers\HomeController.cs.   Happy to answer any questions you may have if you are not familiar with MVC or whatever.

**Comment by eduardo-mior** [comment]
@jrichview I understand. Good luck, sorry I can't help.

**Comment by RobinHerbots** [comment]
Hi @jrichview ,
 Thx for the demo project!

You can solve the problem by (Index.cshtml)
```
        $("#FromDate").inputmask({
            "mask": "99/99/9999",
            "placeholder": "MM/DD/YYYY"
        });
        $("#FromDate").prop("placeholder", "MM/DD/YYYY");
```

This is what I mentioned before.  (maskplaceholder vs the html placeholder)

The question remaining.  Should it also remove the "maskplaceholder' when submitting with the  "clearMaskOnLostFocus": false.  Maybe someone expects that the masktemplate gets submitted to the server.

**Comment by jrichview** [comment]
Well my opinion is that it should behave analogously to html placeholder because that is probably what most web developers would tend to expect from an input property called "placeholder", right ?

If the placeholder text is submitted as a value,  then it's no longer just a placeholder. It's a mask. 

Placeholder and mask should not be thought of as the same thing. A placeholder displays helpful text to guide the user what is expected for data entry, and this means quite often that it does not match format of what the input will expect.  

For example placeholder might be "Enter a date in month/day/year" format whereas the mask would be "99/99/9999" or whatever.  One cannot assume placeholder text is a valid value for the input field. 

If the plugin gives developers the option of conflating them by submitting placeholders as values, that's fine but it should not be default behavior.

**Comment by jrichview** [comment]
Maybe I didn't address your question clearly. "clearMaskOnLostFocus" obviously refers to the mask which is not the same as placeholder. Placeholder text should not be an input value unless you create some option like "submitPlaceholderAsValue" or some.

**Comment by RobinHerbots** [comment]
@jrichview,

I can follow you but the placeholder option is not a replacement of the html placeholder.  Have a read in the readme. https://github.com/RobinHerbots/Inputmask#placeholder-1

Maybe I should have named it otherwise, to make this more clear.  Anyway, the empty mask template will not be submitted anymore regardless what option is used.

> For example placeholder might be "Enter a date in month/day/year" format whereas the mask would be "99/99/9999" or whatever. One cannot assume placeholder text is a valid value for the input field.

This is behavior for the html placeholder and cannot be achieved with the placeholder option of the inputmask as there is a clear difference in function.

**Comment by jrichview** [comment]
I do understand now, thank you Robin. 

In this particular case my misunderstanding was caused by the fact that I didn't originally add this plugin to our project. I just saw it being used and made the assumption what "placeholder" option meant. Obviously that is largely on me because I didn't research that option in the documentation. But the reason I didn't research is I thought I understood it.

I see now that you are working on a new release. I will give it a try. 

Thank you for hearing me out, and also thanks for the great work on the plugin. I don't know if I can get my client to donate something in return for using the plugin, but if not I will donate something when we get close to release.
